app/snaputils.cpp

```diff
@@ -0,0 +1,203 @@
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "snaputils.h"
+
+SnapUtils::SnapUtils( QObject *parent )
+  : QObject{parent}
+{
+}
+
+
+QgsProject *SnapUtils::qgsProject() const
+{
+  return mQgsProject;
+}
+
+void SnapUtils::setQgsProject( QgsProject *newQgsProject )
+{
+  if ( mQgsProject == newQgsProject )
+    return;
+  mQgsProject = newQgsProject;
+  emit qgsProjectChanged( mQgsProject );
+
+  setupSnapping();
+}
+
+QgsQuickMapSettings *SnapUtils::mapSettings() const
+{
+  return mMapSettings;
+}
+
+void SnapUtils::setMapSettings( QgsQuickMapSettings *newMapSettings )
+{
+  if ( mMapSettings == newMapSettings )
+    return;
+
+  if ( mMapSettings )
+  {
+    disconnect( mMapSettings, &QgsQuickMapSettings::extentChanged, this, &SnapUtils::setup );
+    disconnect( mMapSettings, &QgsQuickMapSettings::destinationCrsChanged, this, &SnapUtils::setup );
+    disconnect( mMapSettings, &QgsQuickMapSettings::mapUnitsPerPixelChanged, this, &SnapUtils::setup );
+    disconnect( mMapSettings, &QgsQuickMapSettings::visibleExtentChanged, this, &SnapUtils::setup );
+    disconnect( mMapSettings, &QgsQuickMapSettings::outputSizeChanged, this, &SnapUtils::setup );
+    disconnect( mMapSettings, &QgsQuickMapSettings::outputDpiChanged, this, &SnapUtils::setup );
```

Comment:
We can simplify the code thanks to the handy [QT disconnect syntax ](https://doc.qt.io/qt-5/qobject.html#disconnect): 

```suggestion
    disconnect( mMapSettings, nullptr, this, nullptr );
```

It should disconnect all connections between `mMapSettings` and `this` receiver.